### ISSUE TITLE
Confusing error message when calling getParameters with inconsistent types

### ISSUE THREAD

**Issue by gdesjardins** [torch/torch7#215]
getParameters() does not explicitly check that the datatypes of all parameters (and gradients) are consistent with one another. This yields a very confusing error message somewhere down the line, as shown below.

```
require 'nn'
local mod = nn.Sequential()
mod:add(nn.Linear(10,20):float())
mod:add(nn.Linear(20,10))
local params, gparams = mod:getParameters()
```

which outputs 

```
$ torch delme.lua
luajit: /Users/gdesjardins/usr/torch/share/lua/5.1/nn/Module.lua:201: bad argument #1 to 'set' (expecting number or Tensor or Storage)
stack traceback:
	[C]: in function 'set'
	/Users/gdesjardins/usr/torch/share/lua/5.1/nn/Module.lua:201: in function 'flatten'
	/Users/gdesjardins/usr/torch/share/lua/5.1/nn/Module.lua:243: in function 'getParameters'
	delme.lua:5: in main chunk
	[C]: at 0x010e025eb0
```

**Comment by soumith** [comment]
Fixed by tim in #245
